Add ModernVBERT models

paultltc opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #42503 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

### Multimodal Model Addition Checklist

Please ensure your PR completes all following items. See the [full checklist](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#vision-language-model-contribution-checklist) for details.

- [X] **Modular file**: `modular_<model_name>.py` implemented and verified with `python utils/modular_model_converter.py <model_name>`
- [X] **Fast image processor**: Implemented using `BaseImageProcessorFast` (see [#36978](https://github.com/huggingface/transformers/issues/36978))
- [ ] **Conversion script**: `convert_<model_name>_to_hf.py` added with usage examples
- [X] **Integration tests**: End-to-end tests with exact output matching (text or logits)
- [X] **Documentation**: Model docs added/updated in `docs/source/en/model_doc/`
- [X] **Pattern reuse**: Verified against similar models (LLaVA, Idefics2, etc.)
- [X] **Quality checks**: `make fixup` passes with no errors


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yonigozlan here it is!

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker @Cyrilvallez
- vision models: @yonigozlan @molbap
- audio models: @eustlb @ebezzam @vasqu
- multimodal models: @zucchini-nlp
- graph models: @clefourrier

Library:

- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- continuous batching: @remi-or @ArthurZucker @McPatate
- pipelines: @Rocketknight1
- tokenizers: @ArthurZucker and @itazap
- trainer: @SunMarc
- attention: @vasqu @ArthurZucker @CyrilVallez
- model loading (from pretrained, etc): @CyrilVallez
- distributed: @3outeille @ArthurZucker
- CIs: @ydshieh

Integrations:

- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization: @SunMarc @MekkCyber
- kernels: @MekkCyber @drbh
- peft: @BenjaminBossan @githubnemo

Devices/Backends:

- AMD ROCm: @ivarflakstad
- Intel XPU: @IlyasMoutawwakil
- Ascend NPU: @ivarflakstad 

Documentation: @stevhliu

Research projects are not maintained and should be taken as is.

 -->

paultltc committed:
modernbert modeling

paultltc committed:
modernbert modeling

paultltc committed:
init back

paultltc committed:
init back

paultltc committed:
add mdoernvbert

paultltc committed:
add mdoernvbert

paultltc committed:
refine test: testing logits

paultltc committed:
refine test: testing logits

paultltc committed:
refine test: testing logits

paultltc committed:
refine test: testing logits

paultltc committed:
add missing methods to modernvbert + clean code and doc

paultltc committed:
add missing methods to modernvbert + clean code and doc

paultltc committed:
add modernvbert and colmodernvbert

paultltc committed:
add modernvbert and colmodernvbert

paultltc committed:
clean

paultltc committed:
clean

paultltc committed:
add auto mapping for colmvb

paultltc committed:
add auto mapping for colmvb

paultltc committed:
colmvb imports

paultltc committed:
colmvb imports

paultltc committed:
add colmvb tests

paultltc committed:
add colmvb tests

paultltc committed:
add mvb and colmvb

paultltc committed:
add mvb and colmvb

paultltc committed:
Merge branch 'add-modernvbert' of https://github.com/paultltc/transformers into add-modernvbert

paultltc committed:
update models for latest transformers version

paultltc committed:
update tests

paultltc committed:
update doc

paultltc committed:
update docs

paultltc committed:
update mvb doc

paultltc committed:
reformat

paultltc committed:
update model name on colmvb test

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+    # Copied from transformers.models.idefics2.modeling_idefics2.Idefics2ForConditionalGeneration.disable_input_require_grads
+    def disable_input_require_grads(self):
+        self._text_require_grads_hook.remove()
+        self._vision_require_grads_hook.remove()
```
This is not in idefics2 (anymore at least), + is it not correctly handled by the base PreTrainedModel?

yonigozlan replied:
These are not in idefics2/smolvlm anymore, and shouldn't be needed. Make sure you pull the up-to-date main branch here to have the latest updates

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+                The attention mask indicating padded regions in the image.
+        """
+        batch_size, num_images, num_channels, height, width = pixel_values.shape
+        # pixel_values = pixel_values.to(dtype=self.dtype)  # fp16 compatibility
```
```suggestion
```

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            in a chat into a tokenizable string.
+    """
+
+    image_processor_class = "ModernVBertImageProcessor"
```
No need to have this anymore

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            int(height / pixel_shuffle_factor),
+            embed_dim * (pixel_shuffle_factor**2),
+        )
+        x = x.permute(0, 2, 1, 3)
```
Let's use more explicit var names here (batch_size, seq_length)

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            "in": self.config.initializer_range,
+            "out": self.config.initializer_range / math.sqrt(2.0 * self.config.num_hidden_layers),
+            "embedding": self.config.initializer_range,
+            "final_out": self.config.hidden_size**-0.5,
```
let's define them as separate vars, not in a dict

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            init_weight(module.dense, stds["out"])
+        elif isinstance(module, ModernVBertForMaskedLM):
+            init_weight(module.lm_head, stds["out"])
+        elif isinstance(module, ModernVBertPredictionHead):
```
These can be grouped?

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+    [*ModernVBERT: Towards Smaller Visual Document Retrievers*](https://arxiv.org/abs/2510.01149).
+    """
+)
+class ModernVBertModel(ModernVBertPreTrainedModel):
```
Let's inherit from `SmolVLMModel` so that we don't have to repeat the methods that are exactly the same. Also it's a way to signal that the model is similar to the inherited model.

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
```
Just saw that this hasn't been updated in smolvlm, but let's update it here.
We don't need `output_attentions`, `output_hidden_states`, return_dict anymore:

```suggestion
        **kwargs: Unpack[TransformersKwargs],
```

TransformersKwargs is imported from ...utils

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
```
```suggestion
```

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            image_hidden_states=image_hidden_states,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
```
```suggestion
```

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+            image_hidden_states=image_hidden_states,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
```
```suggestion
```

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1224 @@
+
+        if not return_dict:
+            output = (logits,)
+            return ((loss,) + output) if loss is not None else output
```
```suggestion
```

yonigozlan commented on src/transformers/modeling_utils.py:
```diff
@@ -185,6 +185,28 @@ def is_local_dist_rank_0():
+    "videollava",
+    "vipllava",
+]
+
```
```suggestion
```

yonigozlan commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,548 @@
+                std=std,
+                a=-cutoff_factor * std,
+                b=cutoff_factor * std,
+            )
```
No need to define a method for this here

yonigozlan reviewed (commented):
Hey @paultltc thanks for working on this and sorry for the delay! Most of my comments are related to making better use of modular ([here's some doc](https://huggingface.co/docs/transformers/en/modular_transformers) in case you missed it).
Also make sure you're up-to-date with the main branch. Other than that, the PR is looking good and should not be too far from ready to be merged!

paultltc commented:
Hey @yonigozlan, thanks for checking this! I'll do the changes as soon as possible!

paultltc commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,548 @@
+
+        Query augmentation buffers are used as reasoning buffers during inference.
+        """
+        return self.end_of_utterance_token
```
I followed the way it is implemented in `ColPaliProcessor`, but I had the same doubt. Maybe we should modify this

paultltc committed:
wip

paultltc committed:
wip

paultltc committed:
Merge remote-tracking branch 'upstream/main' into add-modernvbert

paultltc commented on src/transformers/models/modernvbert/processing_modernvbert.py:
```diff
@@ -0,0 +1,41 @@
+#                🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
```
yes I used this command. what is the issue here? this might come from the `DEFAULT_CHAT_TEMPLATE` that I was using before

paultltc committed:
merge main

paultltc committed:
fix test

paultltc commented:
@yonigozlan I updated the code with your suggestions! 

Some might still be discussed:
- `ModernVBertModel` not inheriting from `SmolVLMModel`, as I believe it would be hard to make it work given the architecture differences (i.e. we would inherit and redefine `__init__` and most of `__call__`). But if you have a clear idea I would be happy to modify this!
- `ColModernVBertProcessor` not inheriting from `ColQwenProcessor`, as I think this would end up in more LoC than inheriting from `ModernVBertProcessor` because of overwritting `__call__`
- I updated `ColQwen2Config` to make `ColModernVBertConfig` really compact following your review.

Happy holidays!

yonigozlan commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,497 @@
+from dataclasses import dataclass
```
Same here for the copyright header

yonigozlan replied:
You are missing the copyright header here

yonigozlan commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,497 @@
+from ...tokenization_utils_base import PreTokenizedInput, TextInput
+from ...utils import ModelOutput, auto_docstring, can_return_tuple, is_torch_available, logging
+from ..colqwen2 import ColQwen2Config
+from ..modernvbert import ModernVBertConfig, ModernVBertModel, ModernVBertProcessor
```
```suggestion
from ..colqwen2.configuration_colqwen2 import ColQwen2Config
from ..modernvbert.configuration import ModernVBertConfig
etc.
```

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -137,6 +123,11 @@ def __init__(
+            logger.warning(
+                "The vocab_size parameter is deprecated, please set vocab_size in the `text_config` instead."
+            )
+
```
No need for that as we haven't merged the model yet

yonigozlan commented on src/transformers/models/colqwen2/configuration_colqwen2.py:
```diff
@@ -54,7 +55,7 @@ class ColQwen2Config(PreTrainedConfig):
 
     model_type = "colqwen2"
-    sub_configs: dict[str, Any] = {"vlm_config": PreTrainedConfig}
+    sub_configs: dict[str, Any] = {"vlm_config": Qwen2VLConfig}
```
I'd rather keep PreTrainedConfig here, as vlm_config can be a `Qwen2_5_VLConfig` as well

yonigozlan commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1053 @@
+import math
```
Same here for the copyright header

yonigozlan commented on utils/check_docstrings.py:
```diff
@@ -189,6 +189,7 @@ class DecoratedItem:
     "CanineTokenizer",
     "ChineseCLIPTextModel",
     "ClapTextConfig",
+    "ColModernVBertProcessor",
```
We shouldn't need to add this here or ModernVBertConfig below. Having modular work correctly should fix this

yonigozlan reviewed (commented):
Hello @paultltc ! Thanks for iterating on this. It seems like there are still some issues with modular. Inheritance from other models shouldn't appear in the modeling files outside the modular files ([see the modular doc](https://huggingface.co/docs/transformers/en/modular_transformers)).

yonigozlan commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,548 @@
+        super().__init__(**kwargs)
+
+    def get_text_config(self, *args, **kwargs) -> PreTrainedConfig:
+        return self.vlm_config.get_text_config(*args, **kwargs)
```
Given my last comment, you might have to override __init__ here

paultltc commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,548 @@
+    }
+
+
+class ColModernVBertProcessor(ModernVBertProcessor):
```
> I'm not sure I understand, it looks like you're not using anything from ModernVBertProcessor / Idefics3Processor, as you override __init__ and __call__. 

I do use the `__call__` method from `super()` (l.187 and 215). Do you think it is worth switching the inheritance to `ColPaliProcessor` and rewrite the Idefics3Processor's __call__ logic?

> Where else would you put it? Also it's better not too change the API for backward compatibility.

I was thinking about a utils method, but you are right let's keep it like this to not break backward compatibility! It was rather out of curiosity :)

paultltc committed:
remove useless

paultltc committed:
fix for modular

paultltc committed:
fix tests

paultltc committed:
modeling

paultltc committed:
modular

paultltc committed:
Merge branch 'main' into add-modernvbert

paultltc committed:
lint

paultltc committed:
lint

paultltc committed:
run style and fix-repo

paultltc committed:
fix test

paultltc committed:
doccccstring

paultltc committed:
fix docstring

paultltc committed:
exclude private modules

paultltc committed:
lint

paultltc committed:
fix attribute bug

paultltc committed:
fix tests: tie text config to upper level for some tests

yonigozlan committed:
Fix duplicated ModernVBertConfig

yonigozlan reviewed (approved):
Hello @paultltc ! Thanks for iterating on this.
It's mostly good to go on my side. A lot of the models we inherit from in this PR are quite outdated, which make the modeling files not quite up to the latest transformers standards, but I'd say this is out of scope for this PR. I'll make a note to refactor these but I don't think we need to wait on these updates to merge this.

Could you have a look @vasqu @ArthurZucker @Cyrilvallez ?

yonigozlan committed:
fixup

vasqu reviewed (commented):
Just some initial review, I think we can simplify a lot more or at least work better with our inheritances --> colmodernvbert processor (idefics 3), modernvbert text model (auto model)

paultltc commented on docs/source/en/model_doc/colmodernvbert.md:
```diff
@@ -0,0 +1,89 @@
+<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```
nostalgia I guess :)

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1236 @@
+    The ModernVBert Model with a multiple choice classification head on top (a linear layer on top of the pooled output and a softmax) e.g. for RocStories/SWAG tasks.
+    """
+)
+class ModernVBertForMultipleChoice(ModernVBertPreTrainedModel):
```
Ok, I'll keep only token and sequence classification as I feel they can be useful

paultltc commented on src/transformers/models/auto/image_processing_auto.py:
```diff
@@ -154,6 +154,7 @@
             ("mobilenet_v2", ("MobileNetV2ImageProcessor", "MobileNetV2ImageProcessorFast")),
             ("mobilevit", ("MobileViTImageProcessor", "MobileViTImageProcessorFast")),
             ("mobilevitv2", ("MobileViTImageProcessor", "MobileViTImageProcessorFast")),
+            ("modernvbert", ("ModernVBertImageProcessor", "ModernVBertImageProcessorFast")),
```
You mean completely removing the slow processor in the modular, or just here?

paultltc commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,582 @@
+                images=images,
+                images_kwargs=output_kwargs["images_kwargs"],
+                text_kwargs=output_kwargs["text_kwargs"],
+            )
```
Again, was using concurrent implementations. but happy to modify! @yonigozlan what do you think about this?

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1236 @@
+        return self.text_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.text_model.set_input_embeddings(value)
```
Would it call the `self.text_model.[method](...)` if I remove it?

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1236 @@
+        "ModernVBertMultiheadAttentionPoolingHead",
+        "ModernVBertEmbeddings",
+        "ModernVBertEncoderLayer",
+    ]
```
how would it work if we go for AutoModel for the text model? I won't have access to the modules anymore right?

vasqu replied:
Modernbert has been refactored to follow standard implementations again (no manual padding/unpadding you've seen), see #43030

So it has its own flags now too https://github.com/huggingface/transformers/blob/e4b22baa9b14560dc37657bd0109563a9bf3a4e3/src/transformers/models/modernbert/modeling_modernbert.py#L346-L359

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,1236 @@
+            logits=logits.float(),
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            image_hidden_states=outputs.image_hidden_states,
```
I didn't get the point sorry

paultltc committed:
changes

paultltc committed:
fix resize error

paultltc committed:
Merge remote-tracking branch 'origin/main' into add-modernvbert

paultltc committed:
fix outputs keys problems + attn_type setting in tests

paultltc committed:
run fix-repo

paultltc committed:
clean: remove doc on ModernVBertProcessor which have been removed

vasqu commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,425 @@
+            chat_template=chat_template,
+            image_seq_len=image_seq_len,
+            **kwargs,
+        )
```
```suggestion
        super().__init__(
            image_processor,
            tokenizer,
            chat_template=None,  # ColModernVBert does not use chat templates
            image_seq_len=image_seq_len,
            **kwargs,
        )
```
just for simplicity

vasqu replied:
Gotcha no worries

vasqu commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,425 @@
+    ) -> BatchFeature:
+        """
+        Prepare for the model one or several texts. This method is a wrapper around the `__call__` method of the ColQwen2Processor's
+        [`ColQwen2Processor.__call__`].
```
Same here re name

vasqu commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
````diff
@@ -0,0 +1,425 @@
+    Example:
+
+    ```python
+    from transformers.models.ColModernVBert import ColModernVBertConfig, ColModernVBertForRetrieval
````
Simple transformers should suffice

vasqu commented on src/transformers/models/modernvbert/modular_modernvbert.py:
````diff
@@ -0,0 +1,664 @@
+
+    Example:
+    ```python
+    >>> from modernvbert import ModernVBertConfig
````
Same for the model, check the init file or __all__ if it doesn't work

vasqu commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,664 @@
+                batch_size, seq_len = inputs_embeds.shape[:2]
+            else:
+                batch_size, seq_len = input_ids.shape[:2]
+        device = input_ids.device if input_ids is not None else inputs_embeds.device
```
Let's always infer this ourselves

vasqu commented on tests/models/colmodernvbert/test_modeling_colmodernvbert.py:
```diff
@@ -0,0 +1,325 @@
+            dtype=scores.dtype,
+        )
+
+        assert torch.allclose(scores, expected_scores, atol=1e-2), f"Expected scores {expected_scores}, got {scores}"
```
let's use self.assert... instead (above as well)

vasqu commented on tests/models/modernvbert/test_modeling_modernvbert.py:
```diff
@@ -0,0 +1,671 @@
+    test_mismatched_shapes = False
+    test_cpu_offload = False  # Disabled due to nn.MultiheadAttention compatibility issues with accelerate
+    test_disk_offload_bin = False  # Disabled due to nn.MultiheadAttention compatibility issues with accelerate
+    test_disk_offload_safetensors = False  # Disabled due to nn.MultiheadAttention compatibility issues with accelerate
```
That's due to siglip i guess?

paultltc replied:
but this vision head probe is the reason why `test_training_gradient_checkpointing` are failing. is there a fix? the easiest I see is to set `"vision_use_head": False` in the test config, but this does not solve the underlying problem

vasqu reviewed (commented):
Already looking much better! I added a lot of smaller comments / nits. The overall gist is solid now tho!

Re can record attention outputs etc: Can you sync with main first? If it doesn't work, I can also take a closer look

paultltc commented on tests/models/modernvbert/test_modeling_modernvbert.py:
```diff
@@ -0,0 +1,671 @@
+
+
+@require_torch
+class ModernVBertForMaskedLMIntegrationTest(unittest.TestCase):
```
should I remove the class?

paultltc committed:
clean check utils

paultltc committed:
Merge remote-tracking branch 'origin/main' into add-modernvbert

paultltc commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,425 @@
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            **kwargs,
+        )
```
it might also be that I'm missing something haha. but removing these lines and keeping everything in kwargs did not work for this class

zucchini-nlp replied:
I also noticed that with T5Gemma, which is nested twice. Tho it should work fine, I fixed the tests to recursively set the value in subconfigs

Prev it was setting only one the first level of sub-nesting

paultltc commented on tests/models/colmodernvbert/test_modeling_colmodernvbert.py:
```diff
@@ -0,0 +1,325 @@
+        )
+        # ModernBERT defaults to FlashAttention which only supports fp16/bf16.
+        # Force SDPA on the text backbone for fp32 test compatibility.
+        config.vlm_config.text_config._attn_implementation = "sdpa"
```
perfect!

paultltc committed:
fixing

paultltc committed:
clean tests

vasqu committed:
Merge branch 'main' into add-modernvbert

vasqu committed:
fix the test for outputs xxx

vasqu committed:
fixup auto

vasqu committed:
more auto cleanup

zucchini-nlp commented on src/transformers/models/colmodernvbert/modular_colmodernvbert.py:
```diff
@@ -0,0 +1,418 @@
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> ColModernVBertForRetrievalOutput:
+        if pixel_values is not None:
+            pixel_values = pixel_values.to(dtype=self.dtype)
```
`self.dtype` could be different from the vision tower's dtype/ I think it's better to let the VLM handle it, we usually cast pixels to correct dtype in `patch_embedding` layer

vasqu reviewed (commented):
> (minor) The visual model probe head breaks somes tests (like originially offload but fixed, gradient_checkpointing). Should we do something?

No worries, I think there have been a few like these sadly 😓 if you find a solid fix I'm all open for it, otherwise fine to skip for now

I've added a few smaller comments but imo these are all minor things and we have the biggest issues behind us 🙌

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,654 @@
+        "ModernBertEncoderLayer",
+        "SiglipEncoderLayer",
+        "SiglipMultiheadAttentionPoolingHead",
+    ]
```
probably not anymore, that is when I tried to solve offload problems. let me remove this

paultltc commented on src/transformers/models/modernvbert/modular_modernvbert.py:
```diff
@@ -0,0 +1,654 @@
+        "SiglipMultiheadAttentionPoolingHead",
+    ]
+    _supports_flex_attn = False
+    _can_record_outputs = {"image_hidden_states": ModernVBertConnector}
```
This is the only one failing tho, all other still pass

vasqu committed:
fix retain grads test

vasqu committed:
fixup skip condition

paultltc committed:
updates

paultltc committed:
Merge branch 'main' into add-modernvbert

vasqu committed:
fix new decorators

vasqu commented on src/transformers/models/colqwen2/configuration_colqwen2.py:
```diff
@@ -19,6 +19,7 @@
 from ...configuration_utils import PreTrainedConfig
 from ...utils import logging
 from ..auto import CONFIG_MAPPING
+from ..qwen2_vl.configuration_qwen2_vl import Qwen2VLConfig  # noqa: F401
```
```suggestion
```

vasqu committed:
last nits

vasqu committed:
style

vasqu reviewed (approved):
One last question, I fixed some nits (+ some new stuff on main)

I will be running the slow tests in a second to check whether everything runs as expected - ~not sure on which gpu you ran your integration tests so might need to change values from our side; just need to confirm that it works as expected currently~ looks good!

vasqu committed:
that's on me

vasqu committed:
decorators are confusing lol

paultltc committed:
remove get_num_multimodal_tokens test

vasqu committed:
fix dates

vasqu committed:
Merge branch 'main' into add-modernvbert

vasqu reviewed (approved):
Thanks a lot for iterating, skipping that one processing test seems reasonable to me.

Waiting for the CI on main to get stable again and then merging 🤗

vasqu committed:
Merge branch 'main' into add-modernvbert
